src/Microsoft.AspNetCore.Server.Kestrel/Http/MessageBody.cs

```diff
@@ -255,8 +267,36 @@ public override ValueTask<int> ReadAsyncImplementation(ArraySegment<byte> buffer
                         }
                         else
                         {
-                            throw new NotImplementedException("INVALID REQUEST FORMAT");
+                            ThrowInvalidFormat();
+                        }
+                    }
+                }
+                while (_mode == Mode.Trailer)
+                {
+                    // Consume request terminator
+                    var scan = input.ConsumingStart();
+                    var consumed = scan;
+                    var ch1 = scan.Take();
+                    var ch2 = scan.Take();
+                    if (ch1 == -1 || ch2 == -1)
+                    {
+                        input.ConsumingComplete(consumed, scan);
+                        if (input.RemoteIntakeFin)
+                        {
+                            ThrowChunkedRequestIncomplete();
                         }
+
+                        await input;
+                    }
+                    else if (ch1 == '\r' && ch2 == '\n')
+                    {
+                        input.ConsumingComplete(scan, scan);
```

Comment:
`ConsumingComplete` should be called in a finally block. We're not consistent about doing that in this class (I'll fix that soon), but that is the recommended calling pattern.